src/JustEat.StatsD/PooledUdpTransport.cs

```diff
@@ -40,66 +99,58 @@ public void Send(string metric)
 
             var bytes = Encoding.UTF8.GetBytes(metric);
             var endpoint = _endpointSource.GetEndpoint();
+            var pool = GetPool(endpoint);
+            var socket = pool.Pool.PopOrCreate();
 
-            var socket = _socketPool.PopOrCreate();
             try
             {
-                socket.SendTo(bytes, endpoint);
+                socket.Send(bytes);
             }
             catch (Exception)
             {
                 socket.Dispose();
                 throw;
             }
 
-            _socketPool.Push(socket);
+            pool.Pool.Push(socket);
         }
-
-        /// <summary>
-        /// Finalizes an instance of the <see cref="PooledUdpTransport"/> class.
-        /// </summary>
-        ~PooledUdpTransport()
-        {
-            Dispose(false);
-        }
-
+        
         /// <inheritdoc />
         public void Dispose()
         {
-            Dispose(true);
-            GC.SuppressFinalize(this);
+            _pool?.Dispose();
         }
 
-        /// <summary>
-        /// Performs application-defined tasks associated with freeing, releasing, or resetting unmanaged resources.
-        /// </summary>
-        /// <param name="disposing">
-        /// <see langword="true" /> to release both managed and unmanaged resources;
-        /// <see langword="false" /> to release only unmanaged resources.
-        /// </param>
-        private void Dispose(bool disposing)
+        private readonly object _lock = new object();
+
+        private ConnectedPool GetPool(IPEndPoint endPoint)
         {
-            if (!_disposed)
+            if (_pool == null)
             {
-                try
+                lock (_lock)
                 {
-                    while (_socketPool.Count > 0)
+                    if (_pool == null)
                     {
-                        var socket = _socketPool.Pop();
-                        socket?.Dispose();
+                        _pool = new ConnectedPool(endPoint);
+                        return _pool;
                     }
                 }
-                catch (ObjectDisposedException)
+            }
+
+            if (endPoint.Equals(_pool.IpEndPoint))
+            {
+                return _pool;
+            }
+            else
+            {
+                lock (_lock)
```

Comment:
I don't think this lock is a good idea, as it's effectively putting a lock around network calls to `Connect()` in the constructor of the pool.